rust-stream/src/main.rs

```diff
@@ -0,0 +1,413 @@
+use std::fmt::{Debug, Display};
+use std::iter::Sum;
+use std::mem::size_of;
+use std::time::{Duration, Instant};
+
+use num_traits::{abs, NumAssign, Signed};
+use num_traits::real::Real;
+use rayon::prelude::*;
+use structopt::StructOpt;
+use tabular::{Row, Table};
+
+#[derive(Debug, StructOpt)]
+struct Options {
+    /// List available devices
+    #[structopt(long)] list: bool,
+    /// Select device at <device>
+    #[structopt(long, default_value = "0")] device: usize,
+    /// Run the test <numtimes> times (NUM >= 2)
+    #[structopt(long, default_value = "100")] numtimes: usize,
+    /// Use <arraysize> elements in the array
+    #[structopt(long, default_value = "33554432")] arraysize: usize,
+    /// Use floats (rather than doubles)
+    #[structopt(long)] float: bool,
+    /// Only run triad
+    #[structopt(long)] triad_only: bool,
+    /// Only run nstream
+    #[structopt(long)] nstream_only: bool,
+    /// Output as csv table
+    #[structopt(long)] csv: bool,
+    /// Use MiB=2^20 for bandwidth calculation (default MB=10^6)
+    #[structopt(long)] mibibytes: bool,
+}
+
+#[derive(PartialEq)]
+enum Benchmark { All, Triad, NStream }
+
+struct StreamData<T> {
+    size: usize,
+    scalar: T,
+    a: Vec<T>,
+    b: Vec<T>,
+    c: Vec<T>,
+}
+
+impl<T: Default + Clone> StreamData<T> {
+    pub fn new(size: usize, scalar: T) -> StreamData<T> {
+        StreamData {
+            size,
+            scalar,
+            a: vec![T::default(); size],
+            b: vec![T::default(); size],
+            c: vec![T::default(); size],
+        }
+    }
+}
+
+struct PlainFor;
+
+struct RayonPar;
+
+#[inline(always)]
+fn timed<F: FnOnce()>(f: F) -> Duration {
+    let start = Instant::now();
+    f();
+    start.elapsed()
+}
+
+#[inline(always)]
+fn timed_mut<T, F: FnMut() -> T>(f: &mut F) -> (Duration, T) {
+    let start = Instant::now();
+    let x = f();
+    (start.elapsed(), x)
+}
+
+struct AllTiming<T> { copy: T, mul: T, add: T, triad: T, dot: T }
+
+trait RustStream<T: Default, K> {
+    fn init_arrays(&mut self, init: (T, T, T));
+    fn copy(&mut self);
+    fn mul(&mut self);
+    fn add(&mut self);
+    fn triad(&mut self);
+    fn nstream(&mut self);
+    fn dot(&mut self) -> T;
```

Comment:
It would be useful to have more comments, for example on these traits. What is the overall goal of the code,
what each of these tests accomplishes, what is the expected behaviour?